src/huggingface_hub/utils/cache_explorer.py

```diff
@@ -0,0 +1,246 @@
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, List, Literal, Optional, Set, Union
+
+from ..constants import HUGGINGFACE_HUB_CACHE
+
+
+@dataclass
+class CachedFileInfo:
+    """Information about a single file.
+
+    The file path is a symlink existing in the `snapshots` folder and referring to a
+    blob in the `blobs` folder.
+    """
+
+    file_name: str
+    file_path: Path
+    size_on_disk: int
+    blob_path: Path
+
+    @property
+    def size_on_disk_str(self) -> str:
+        return _format_size(self.size_on_disk)
+
+
+@dataclass
+class CachedRevisionInfo:
+    r"""Information about a revision (a snapshot).
+
+    A revision can be either referenced by 1 or more `refs` or be "detached" (no refs).
+
+    /!\ `size_on_disk` is not necesarily the sum of all file sizes because of possible
+        duplicated files. Also only blobs are taken into account, not the (neglectable)
+        size of folders and symlinks.
+    """
+
+    commit_hash: str
+    size_on_disk: int
+    files: List[CachedFileInfo]  # sorted by name
+    refs: Optional[Set[str]] = None
+
+    @property
+    def size_on_disk_str(self) -> str:
+        return _format_size(self.size_on_disk)
+
+    @property
+    def nb_files(self) -> int:
+        return len(self.files)
+
+
+@dataclass
+class CachedRepoInfo:
+    r"""Information about a cached repository (dataset or model).
+
+    /!\ `size_on_disk` is not necesarily the sum of all revisions sizes because of
+        duplicated files. Also only blobs are taken into account, not the (neglectable)
+        size of folders and symlinks.
+    """
+
+    repo_id: str
+    repo_type: Literal["model", "dataset"]
+    repo_path: Path
+    revisions: Set[CachedRevisionInfo]
+    size_on_disk: int
+    nb_files: int
+
+    @property
+    def size_on_disk_str(self) -> str:
+        """Human-readable sizes"""
+        return _format_size(self.size_on_disk)
+
+
+@dataclass
+class HFCacheInfo:
+    r"""Information about the entire cache repository.
+
+    /!\ Here `size_on_disk` is equal to the sum of all repo sizes (only blobs).
+    """
+
+    datasets: List[CachedRepoInfo]
+    models: List[CachedRepoInfo]
+    size_on_disk: int
+
+    @property
+    def size_on_disk_str(self) -> str:
+        return _format_size(self.size_on_disk)
+
+
+class CorruptedCacheException(Exception):
+    """Exception for any unexpected structure in the huggingface cache."""
+
+
+def scan_cache_dir(
+    cache_dir: Optional[Union["str", Path]] = None
+) -> List[CachedRepoInfo]:
+    """Scan the entire cache directory and return information about it."""
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+
+    datasets: List[CachedRepoInfo] = []
+    models: List[CachedRepoInfo] = []
```

Comment:
Type annotations in signatures are fine, but in code I often find they make it harder to read than really illuminate things (especially when variables are aptly named like here).

Reply:
personally i like them (but i like types in general unlike @sgugger 🙈 )

Reply:
I'll keep them. Typing doesn't hurt 🙄